### PR TITLE
Add badges pallet

### PR DESCRIPTION
**Pull request by samchuk-vlad**
- [x] Add events
- [x] Add fixes after review
- [ ] Finish TODOs in code

From @siman:

- [ ] create `BadgesSettings` per space: `{ accepting_badges: boolean = true, etc... }`

### CODE REVIEW

File: pallets/badges/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+	trait Store for Module<T: Trait> as TemplateModule {
+        NextBadgeId get(fn next_badge_id): BadgeId = 1;
+        NextSpaceAwardId get(fn next_space_award_id): SpaceAwardId = 1;
+
+        pub BadgeById get(fn badge_by_id):
+            map hasher(twox_64_concat) BadgeId => Option<Badge<T>>;
+
+        pub SpaceAwardById get(fn space_award_by_id):
+            map hasher(twox_64_concat) SpaceAwardId => Option<SpaceAward<T>>;
+
```
Comment (F3Joule) [inline pallets/badges/src/lib.rs:60]:
```suggestion

        pub SpaceAwardsBySpaceId get(fn space_awards_by_space_id):
            map hasher(twox_64_concat) SpaceId => Vec<SpaceAwardId>;
 
```

File: pallets/badges/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    space_id: SpaceId,
+}
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug)]
+pub struct SpaceAward<T: Trait> {
+    created: WhoAndWhen<T>,
+    badge_id: BadgeId,
+    recipient: SpaceId,
+    expires_at: Option<T::BlockNumber>,
+    approved: bool
```
Comment (F3Joule) [inline pallets/badges/src/lib.rs:40]:
```suggestion
    accepted: bool
```

File: pallets/badges/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            SpaceAwardsByBageId::mutate(badge_id, |ids| ids.push(space_award_id));
+
+            <SpaceAwardById<T>>::insert(space_award_id, new_space_award);
+            NextSpaceAwardId::mutate(|x| *x += 1);
+
+            Ok(())
+        }
+
+        #[weight = T::DbWeight::get().reads_writes(2, 1) + 10_000]
+        pub fn approve_space_award(origin, space_award_id: SpaceAwardId) -> DispatchResult {
```
Comment (F3Joule) [inline pallets/badges/src/lib.rs:174]:
```suggestion
        pub fn accept_space_reward(origin, space_award_id: SpaceAwardId) -> DispatchResult {
```

File: pallets/badges/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+                <SpaceAwardById<T>>::remove(id);
+            }
+
+            <BadgeById<T>>::remove(badge_id);
+
+            Ok(())
+        }
+
+        #[weight = T::DbWeight::get().reads_writes(6, 4) + 10_000]
+        pub fn propose_space_award(
```
Comment (F3Joule) [inline pallets/badges/src/lib.rs:139]:
```suggestion
        pub fn award_badge(
```

Comment (siman) [inline pallets/badges/src/lib.rs:139]:
```suggestion
        pub fn award_badge(
```

File: pallets/badges/src/lib.rs (hunk 3 above)
Comment (siman) [inline pallets/badges/src/lib.rs:174]:
```suggestion
        pub fn accept_award(origin, space_award_id: SpaceAwardId) -> DispatchResult {
```

File: pallets/badges/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+            Ok(())
+        }
+
+        #[weight = T::DbWeight::get().reads_writes(1, 1) + 10_000]
+        pub fn delete_space_award(origin, space_award_id: SpaceAwardId) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            let space_award = Self::space_award_by_id(space_award_id).ok_or(Error::<T>::SpaceAwardNotFound)?;
+            let badge = Self::badge_by_id(space_award.badge_id).ok_or(Error::<T>::BadgeNotFound)?;
+            Self::ensure_badge_manager(who, badge.space_id)?;
```
Comment (siman) [inline pallets/badges/src/lib.rs:195]:
```suggestion
            Self::ensure_award_manager(who, badge.space_id)?;
```

File: pallets/badges/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+            space.ensure_space_owner(who)?;
+
+            space_award.approved = true;
+            <SpaceAwardById<T>>::insert(space_award_id, space_award);
+
+            Ok(())
+        }
+
+        #[weight = T::DbWeight::get().reads_writes(1, 1) + 10_000]
+        pub fn delete_space_award(origin, space_award_id: SpaceAwardId) -> DispatchResult {
```
Comment (siman) [inline pallets/badges/src/lib.rs:190]:
```suggestion
        pub fn refuse_award(origin, space_award_id: SpaceAwardId) -> DispatchResult {
```

File: pallets/badges/src/lib.rs (hunk 5 above)
Comment (F3Joule) [inline pallets/badges/src/lib.rs:195]:
```suggestion
            Self::ensure_award_manager(who, badge.space_id)?;
```

File: runtime/src/lib.rs [hunk 7]
```diff
@@ -341,6 +341,9 @@ parameter_types! {
       SP::UpdateEntityStatus,
 
       SP::UpdateSpaceSettings,
+
```
Comment (F3Joule) [inline runtime/src/lib.rs:344]: Update Runtime mockup in `df-integration-tests`

File: pallets/badges/src/lib.rs (hunk 2 above)
Comment (siman) [inline pallets/badges/src/lib.rs:40]:
```suggestion
    accepted: bool
```

File: pallets/badges/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        pub BadgeById get(fn badge_by_id):
+            map hasher(twox_64_concat) BadgeId => Option<Badge<T>>;
+
+        pub SpaceAwardById get(fn space_award_by_id):
+            map hasher(twox_64_concat) SpaceAwardId => Option<SpaceAward<T>>;
+
+        pub SpaceAwardIdByExpirationBlock get(fn space_award_id_by_expiration_block):
+            map hasher(blake2_128_concat) T::BlockNumber => Vec<SpaceAwardId>;
+
+        pub SpaceAwardsByBageId get(fn space_awards_by_badge_id):
```
Comment (siman) [inline pallets/badges/src/lib.rs:64]:
```suggestion
        pub SpaceAwardsByBadgeId get(fn space_awards_by_badge_id):
```

**Review by siman — CHANGES_REQUESTED** [review]

File: pallets/badges/src/lib.rs (hunk 6 above)
Comment (spencerbh) [reply pallets/badges/src/lib.rs:190]:
```suggestion
        pub fn remove_space_award(origin, space_award_id: SpaceAwardId) -> DispatchResult {
```

**Pull request closed without merging**